### ISSUE TITLE
problem when slurping a file in perl 5.6.0

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2138]
Migrated from [rt.perl.org#3432](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3432) (status was 'resolved')

Searchable as RT3432$

**Comment by p5pRT** [comment]
### From mm5830@bna.com

This is a bug report for perl from mm5830@&#8203;aquarius\.bna\.com\,
generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.

I've run into some recent problems when upgrading from perl 5\.004\_05
to 5\.6\.0\.  Here is the first one&#8203;:

The "\-0777n" command\-line option to perl causes duplicate execution\.
I seem to get an extra iteration of the while \(\<>\) loop\.  Also\, the
same thing happens if the effect of "\-0777n" is\, instead\, implemented
within the perl script\.  Here are some examples that compare results
from perl version 5\.00405 and perl version 5\.6\.0&#8203;:


&nbsp;      Example A
&nbsp;      \-\-\-\-\-\-\-\-\-

&nbsp;     /bin/echo "abc" | perl  \-0777ne ' print "111 $\_ 222\\n"'

&nbsp;           5\.00405
&nbsp;           =======> 111 abc
&nbsp;             222

&nbsp;           5\.6\.0
&nbsp;           =======> 111 abc
&nbsp;                     222
&nbsp;                    111  222

Why is \<> returning an empty string instead of an undef the second
time through the loop?  Why is the empty string treated as true by the
while loop?


&nbsp;      Example B
&nbsp;      \-\-\-\-\-\-\-\-\-

&nbsp;         /bin/echo "abc" \\
&nbsp;         | perl \-e 'undef $/; while \(\<>\) \{print "111 $\_ 222\\n"\}'

&nbsp;           5\.00405
&nbsp;           =======> 111 abc
&nbsp;                     222

&nbsp;           5\.6\.0
&nbsp;           =======> 111 abc
&nbsp;                     222
&nbsp;                    111  222


&nbsp;      Example C  \(The problem does NOT occur in this case\)
&nbsp;      \-\-\-\-\-\-\-\-\-

&nbsp;         /bin/echo "abc" \\
&nbsp;         | perl \-e 'undef $/; $\_ = \<>; print "111 $\_ 222\\n"'

&nbsp;           =======> 111 abc
&nbsp;                     222

That's because I am just reading from \<> once\.

Thanks\,
Mike
mikemulvaney@&#8203;excite\.com


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=high


---
Site configuration information for perl v5\.6\.0&#8203;:

Configured by mm5830 at Tue Jun 20 10&#8203;:06&#8203;:34 EDT 2000\.

Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.6\, archname=sun4\-solaris
&nbsp;    uname='sunos aquarius 5\.6 generic\_105181\-11 sun4d sparc sunw\,sparcserver\-1000 '
&nbsp;    config\_args='\-ds \-e \-Dprefix=/devl'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define 
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-O'\, gccversion=
&nbsp;    cppflags=''
&nbsp;    ccflags =' \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    stdchar='unsigned char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, usemymalloc=y\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/test/SUNWspro/SC5\.0/lib '
&nbsp;    libpth=/test/SUNWspro/SC5\.0/lib /lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-ldl \-lm \-lc \-lcrypt \-lsec
&nbsp;    libc=/lib/libc\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-KPIC'\, lddlflags='\-G \-L/test/SUNWspro/SC5\.0/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.6\.0&#8203;:
&nbsp;    /devl/lib/perl5/5\.6\.0/sun4\-solaris
&nbsp;    /devl/lib/perl5/5\.6\.0
&nbsp;    /devl/lib/perl5/site\_perl/5\.6\.0/sun4\-solaris
&nbsp;    /devl/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /devl/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl v5\.6\.0&#8203;:
&nbsp;    HOME=/home/mm5830
&nbsp;    LANG=C
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH=/usr/lib&#8203;:/usr/dt&#8203;:/usr/dt/lib&#8203;:/usr/openwin/lib&#8203;:/devl/lib&#8203;:/devl/ot/lib&#8203;:/opt/lotus/common/lel/r100/sunspa53
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/mm5830/bin&#8203;:/home/mm5830/bin&#8203;:/devl/tools/bin&#8203;:\.&#8203;:/devl/bin&#8203;:/devl/jdk/bin&#8203;:/usr/bin&#8203;:/usr/sbin&#8203;:/usr/openwin/bin&#8203;:/etc&#8203;:/usr/ccs/bin&#8203;:/usr/ccs/lib&#8203;:/devl/ot/bin&#8203;:/devl/adept60/bin&#8203;:/opt/lotus/bin&#8203;:/home/mm5830/notes&#8203;:/devl/voyager/bin&#8203;:/test/informix/bin&#8203;:/test/SoftWindows/bin&#8203;:/devl/frame/bin&#8203;:/test/SUNWspro/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/csh

**Comment by p5pRT** [comment]
### From @gsar

On Tue\, 27 Jun 2000 13&#8203;:31&#8203;:49 EDT\, Michael Mulvaney wrote&#8203;:
>     /bin/echo "abc" | perl  \-0777ne ' print "111 $\_ 222\\n"'
>
>           5\.00405
>           =======> 111 abc
>             222
>
>           5\.6\.0
>           =======> 111 abc
>                     222
>                    111  222
>
>Why is \<> returning an empty string instead of an undef the second
>time through the loop?

Hans Mulder had the right idea about this all along\, but I was apparently
not paying enough attention\.  :\-\(

&nbsp;    http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2000\-01/msg00002\.html

This also covers ID\#20000629\.007 sent in by Paul Grassie\.


Sarathy
gsar@&#8203;ActiveState\.com
<details><summary>Inline Patch</summary>

```diff
-----------------------------------8<-----------------------------------
Change 6294 by gsar@auger on 2000/06/30 14:47:45

	slurp mode fix in change#4736 still not quite right

Affected files ...

... //depot/perl/pp_hot.c#172 edit
... //depot/perl/t/io/argv.t#11 edit

Differences ...

==== //depot/perl/pp_hot.c#172 (text) ====
Index: perl/pp_hot.c
--- perl/pp_hot.c.~1~	Fri Jun 30 07:47:56 2000
+++ perl/pp_hot.c	Fri Jun 30 07:47:56 2000
@@ -1390,8 +1390,7 @@
 /* delay EOF state for a snarfed empty file */
 #define SNARF_EOF(gimme,rs,io,sv) \
     (gimme != G_SCALAR || SvCUR(sv)					\
-     || !RsSNARF(rs) || (IoFLAGS(io) & IOf_NOLINE)			\
-     || ((IoFLAGS(io) |= IOf_NOLINE), FALSE))
+     || (IoFLAGS(io) & IOf_NOLINE) || !RsSNARF(rs))
 
     for (;;) {
 	if (!sv_gets(sv, fp, offset)
@@ -1424,6 +1423,7 @@
 	    SvTAINTED_on(sv);
 	}
 	IoLINES(io)++;
+	IoFLAGS(io) |= IOf_NOLINE;
 	SvSETMAGIC(sv);
 	XPUSHs(sv);
 	if (type == OP_GLOB) {

==== //depot/perl/t/io/argv.t#11 (xtext) ====
Index: perl/t/io/argv.t
--- perl/t/io/argv.t.~1~	Fri Jun 30 07:47:56 2000
+++ perl/t/io/argv.t	Fri Jun 30 07:47:56 2000
@@ -5,7 +5,7 @@
     unshift @INC, '../lib';
 }
 
-print "1..20\n";
+print "1..21\n";
 
 use File::Spec;
 
@@ -107,18 +107,20 @@
     local $/;
     open F, 'Io_argv1.tmp' or die;
     <F>;	# set $. = 1
+    print "not " if defined(<F>); # should hit eof
+    print "ok 16\n";
     open F, $devnull or die;
     print "not " unless defined(<F>);
-    print "ok 16\n";
-    print "not " if defined(<F>);
     print "ok 17\n";
     print "not " if defined(<F>);
     print "ok 18\n";
+    print "not " if defined(<F>);
+    print "ok 19\n";
     open F, $devnull or die;	# restart cycle again
     print "not " unless defined(<F>);
-    print "ok 19\n";
+    print "ok 20\n";
     print "not " if defined(<F>);
-    print "ok 20\n";
+    print "ok 21\n";
     close F;
 }
 
End of Patch.
```
</details>
